cmd/snap/cmd_run.go

```diff
@@ -307,6 +319,21 @@ func (x *cmdRun) snapRunHook(snapName string) error {
 	return x.runSnapConfine(info, hook.SecurityTag(), snapName, hook.Name, nil)
 }
 
+func (x *cmdRun) snapRunTimer(snapApp, timer string, args []string) error {
+	schedule, err := timeutil.ParseSchedule(timer)
+	if err != nil {
+		return fmt.Errorf("invalid timer format: %v", err)
+	}
+
+	now := timeNow()
+	if !timeutil.Includes(schedule, now) {
+		fmt.Fprintf(Stderr, "on %v attempted to run %v timer outside of scheduled time %q\n", now, snapApp, timer)
```

Comment:
Is it possible the schedule is missed by a tiny bit just because of an imprecision of systemd timer, if executed close to the end of the schedule window?

Reply:
Yes. There's a couple of things we try to do to mitigate this problem. For a schedule like `10:00-12:00` the timer is set to run at the start of the window, i.e. at 10:00. For a schedule with spread, eg. `10:00~12:00`, the OnCalendar event time is chosen randomly between the `[start, end - 5 minutes]`times. If the schedule on a single time, eg. `10:00`, there's a 1 minute window introduced in `timeutil.Includes()` as discussed in #4676. We can tweak it if needed. 

systemd will try to start the service when the time matches OnCalendar specification. The timers are set to be `Persistent`, so systemd will also do the bookkeeping.

An alternative approach was to construct OnCalendar such that it matches multiple 'times' within the window, thus causing `snap run --time` to be run multiple times. But that variant would require us to keep the timer state information (i.e. whether the timer ran or not).

Reply:
Ah, so that's been thought of already, excellent! Thanks for explaining!